default/AI/ResearchAI.py

```diff
@@ -453,28 +492,45 @@ def generate_research_orders():
     rng = random.Random()
     rng.seed(fo.getEmpire().name + fo.getGalaxySetupData().seed)
 
-    research_reqs = calculate_research_requirements(empire)
+    calculate_research_requirements(empire)
     total_rp = empire.resourceProduction(fo.resourceType.research)
 
     if total_rp <= 0: # No RP available - no research.
         return
 
+    raw_priorities = {}
     priorities = {}
+    on_path_to = {}
     for tech_name in fo.techs():
-        priority = get_priority(rng, tech_name)
+        this_tech = fo.getTech(tech_name)
+        if not this_tech:
+            continue
+        raw_priorities[tech_name] = get_priority(rng, tech_name)
+    for tech_name, priority in raw_priorities.iteritems():
         if not tech_is_complete(tech_name) and priority >= 0:
-            turn_needed = float(research_reqs[tech_name][1]) / total_rp
-            priorities[tech_name] = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
+            turn_needed = max(research_reqs[tech_name][2], math.ceil(float(research_reqs[tech_name][1]) / total_rp))
+            adjusted_priority = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
```

Comment:
I don't think treating anything below x turns as equivalent is a problem considering that the unweighted cost of the techs will already provide an order.

Concerning time: I am not too sure how a formula would look like to consider time correctly. I wouldn't say that long research times on techs are a bad thing if the RP cost is low. We gotta research it eventually anyway and the later we start, the later we get the tech (eventhough we could afford to research it earlier). Of course if facing the choice of two otherwise equivalent techs, we would prefer the one we get first.

Still, I think time considerations are (complicated) second order effects (I'm talking about the minimum research time here not the ratio of available RP and cost!). Correspondingly, these shouldn't have too much of an impact and I also think that it would be more efficient to tackle other things first that have more impact and also are more obvious.

So my suggestion would be to first ignore the number of turns to finish the techs (beyond prefering aleady started techs).

As far as your first point goes: Even if you manage to implement that many staged tech line it does not solve the problem of a single tech which is 40RP/20 turn. These things is something we want to start *early* if we have the spare RP so we surely do not want to decrease the priority based on the time...